File: binary-search/704_binary_search.py
"""
Problem: Binary Search
LeetCode #: 704
Difficulty: Easy
Link: https://leetcode.com/problems/binary-search/

Approach: Classic binary search on sorted array.
Time Complexity: O(log n)
Space Complexity: O(1)
"""

from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1

        while left <= right:
            mid = left + (right - left) // 2

            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1

        return -1


# --- Test ---
if __name__ == "__main__":
    sol = Solution()
    print(sol.search([-1, 0, 3, 5, 9, 12], 9))   # 4
    print(sol.search([-1, 0, 3, 5, 9, 12], 2))    # -1

File: binary-search/704_binary_search.java
/*
 * Problem: Binary Search
 * LeetCode #: 704
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/binary-search/
 *
 * Approach: Standard binary search with left and right pointers.
 * Time Complexity: O(log n)
 * Space Complexity: O(1)
 */

public class Solution {
    public int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) return mid;
            if (nums[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }

    // --- Test ---
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.search(new int[]{-1, 0, 3, 5, 9, 12}, 9)); // 4
    }
}
